Add documentation for ingest-attachment plugin (#7891)
opensearch-project/documentation-website: _install-and-configure/additional-plugins/ingest-attachment-plugin.md

---
layout: default
title: Ingest-attachment plugin
parent: Installing plugins
nav_order: 20

---

# Ingest-attachment plugin

The Ingest-attachment plugin enables OpenSearch to extract content and other information from files using the Apache text extraction library [Tika](https://tika.apache.org/).
Supported document formats include PPT, PDF, RTF, ODF and many more ([Tika Supported Document Formats](https://tika.apache.org/2.9.2/formats.html)).

The input field must be a base64-encoded binary.

## Installing the plugin

Install the `ingest-attachment` plugin using the following command:

```sh
./bin/opensearch-plugin install ingest-attachment
```

## Attachment processor options

| Name | Required | Default | Description |

> | :--- | :--- | :--- | :--- |

Reviewer: Line below: What does the hyphen (-) in the Default column signify?

Reviewer 2: Doesn't contain a default.

> | `field` | Yes | - | The field from which to get the base64-encoded binary. |

Reviewer 2: ```suggestion
| `field` | Yes | None | The field from which to get the base64-encoded binary. |
```

> | `target_field` | no | attachment | The field that holds the attachment information. |

Reviewer: ```suggestion
| `target_field` | No | Attachment | The field that stores the attachment information. |
```

> | `properties` | no | all properties | An array of properties, which should be stored. Can be `content`, `language`, `date`, `title`, `author`, `keywords`, `content_type`, `content_length`. |

Reviewer: ```suggestion
| `properties` | No | All properties | An array of properties that should be stored. Can be `content`, `language`, `date`, `title`, `author`, `keywords`, `content_type`, or `content_length`. |
```

> | `indexed_chars` | no | `100_000` | The number of character used for extraction to prevent fields from becoming to large. Use `-1` for no limit. |

Reviewer: ```suggestion
| `indexed_chars` | No | `100_000` | The number of characters used for extraction to prevent fields from becoming too large. Use `-1` for no limit. |
```

> | `indexed_chars_field` | no | `null` | The field name from which you can overwrite the number of chars being used for extraction, for example, `indexed_chars`. |

Reviewer: ```suggestion
| `indexed_chars_field` | No | `null` | The field name from which you can overwrite the number of chars being used for extraction, for example, `indexed_chars`. |
```

> | `ignore_missing` | no | `false` | When `true`, the processor exits without modifying the document when the specified field doesn't exist. |

Reviewer: ```suggestion
| `ignore_missing` | No | `false` | When `true`, the processor exits without modifying the document when the specified field doesn't exist. |
```

> 

Reviewer: Line 32: "The field name from which you can overwrite" => "The field name used to overwrite"?

## Example

> The following steps show how to get started with the Ingest-attachment plugin.

Reviewer: ```suggestion
The following steps show you how to get started with the `ingest-attachment` plugin.
```

> 

Reviewer: If the following are sequential steps, should they be numbered?

> ### Create an index to store your attachments

Reviewer: ```suggestion
### Create an index for storing your attachments
```

The following command creates an index for storing your attachments:

```json
PUT /example-attachment-index
{
  "mappings": {
    "properties": {}
  }
}
```

> ### Create a pipeline with attachment processor

Reviewer: ```suggestion
### Create a pipeline with the attachment processor
```

> ### Create a pipeline with attachment processor

Reviewer 2: ```suggestion
### Step 2: Create a pipeline with attachment processor
```

> The following command creates a pipeline which contains the attachment processor:

Reviewer: ```suggestion
The following command creates a pipeline containing the attachment processor:
```

```json
PUT _ingest/pipeline/attachment
{
  "description" : "Extract attachment information",
  "processors" : [
    {
      "attachment" : {
        "field" : "data"
      }
    }
  ]
}
```

> ### Store an attachment

Reviewer 2: ```suggestion
### Step 3: Store an attachment
```

> Convert the attachment to base64 string, to pass it as `data`.

Reviewer: ```suggestion
Convert the attachment to a base64 string to pass it as `data`.
```

> In this example the Unix-like system `base64` command converts the file `lorem.rtf`:

Reviewer: Is "system" necessary here, or can it be safely removed?

Reviewer 2: It's not. Let's remove it.

> In this example the Unix-like system `base64` command converts the file `lorem.rtf`:

Reviewer 2: ```suggestion
In this example the `base64` command converts the file `lorem.rtf`:
```

```sh
base64 lorem.rtf
```

> Alternatively you can use Node.js to read the file to `base64`, as shown in the following commands:

Reviewer: ```suggestion
Alternatively, you can use Node.js to read the file to `base64`, as shown in the following commands:
```

```typescript
import * as fs from "node:fs/promises";
import path from "node:path";

const filePath = path.join(import.meta.dirname, "lorem.rtf");
const base64File = await fs.readFile(filePath, { encoding: "base64" });

console.log(base64File);
```

> The following base64 string is for an `.rtf` file containing the text

Reviewer: Please clarify "the text". What text, specifically? Also, this line should end in a colon.

> The following base64 string is for an `.rtf` file containing the text

Reviewer 2: ```suggestion
The`.rtf` file contains the following base64 text:
```

```json
PUT example-attachment-index/_doc/lorem_rtf?pipeline=attachment
{
  "data": "e1xydGYxXGFuc2kNCkxvcmVtIGlwc3VtIGRvbG9yIHNpdCBhbWV0DQpccGFyIH0="
}
```

### Query results

With the attachment processed, you can now search through the data using search queries, as shown in the following example:

```json
POST example-attachment-index/_search
{
  "query": {
    "match": {
      "attachment.content": "ipsum"
    }
  }
}
```

OpenSearch responds with the following:

```json
{
  "took": 5,
  "timed_out": false,
  "_shards": {
    "total": 1,
    "successful": 1,
    "skipped": 0,
    "failed": 0
  },
  "hits": {
    "total": {
      "value": 1,
      "relation": "eq"
    },
    "max_score": 1.1724279,
    "hits": [
      {
        "_index": "example-attachment-index",
        "_id": "lorem_rtf",
        "_score": 1.1724279,
        "_source": {
          "data": "e1xydGYxXGFuc2kNCkxvcmVtIGlwc3VtIGRvbG9yIHNpdCBhbWV0DQpccGFyIH0=",
          "attachment": {
            "content_type": "application/rtf",
            "language": "pt",
            "content": "Lorem ipsum dolor sit amet",
            "content_length": 28
          }
        }
      }
    ]
  }
}
```

> ## Extracted Information

Reviewer: ```suggestion
## Extracted information
```

The following fields can be extracted using the plugin:

- `content`
- `language`
- `date`
- `title`
- `author`
- `keywords`
- `content_type`
- `content_length`

To extract only a subset of these fields, define them in the `properties` of the

> pipelines processor, as shown in the following example:

Reviewer: ```suggestion
pipeline processor, as shown in the following example:
```

```json
PUT _ingest/pipeline/attachment
{
  "description" : "Extract attachment information",
  "processors" : [
    {
      "attachment" : {
        "field" : "data",
        "properties": ["content", "title", "author"]
      }
    }
  ]
}
```

## Limit the extracted content

> To prevent extracting too many characters and overload the node memory, the default limit is `100_000`.

Reviewer: ```suggestion
To prevent extracting too many characters and overloading the node memory, the default limit is `100_000`.
```

> You can change this value using the setting `indexed_chars`. For example, you can use `-1` for unlimited characters but you need to make sure you have enough HEAP space on your OpenSearch-Node to extract the content of large documents.

Reviewer: ```suggestion
You can change this value using the setting `indexed_chars`. For example, you can use `-1` for unlimited characters, but you need to make sure you have enough HEAP space on your OpenSearch node to extract the content of large documents.
```

You can also define this limit per document using the `indexed_chars_field` request field.
If a document contains `indexed_chars_field`, it will overwrite the `indexed_chars` setting, as shown in the following example:

```json
PUT _ingest/pipeline/attachment
{
  "description" : "Extract attachment information",
  "processors" : [
    {
      "attachment" : {
        "field" : "data",
        "indexed_chars" : 10,
        "indexed_chars_field" : "max_chars",
      }
    }
  ]
}
```

> With the attachment pipeline set, you can extract the above defaulted `10` characters without specifying `max_chars` in the request, as shown in the following example:

Reviewer: ```suggestion
With the attachment pipeline configured, you can extract the default `10` characters without specifying `max_chars` in the request, as shown in the following example:
```

```json
PUT example-attachment-index/_doc/lorem_rtf?pipeline=attachment
{
  "data": "e1xydGYxXGFuc2kNCkxvcmVtIGlwc3VtIGRvbG9yIHNpdCBhbWV0DQpccGFyIH0="
}
```

> Alternatively, you can change the `max_char` per document to extract up to `15` characters, as shown in the following example:

Reviewer: ```suggestion
Alternatively, you can change the `max_char` per document in order to extract up to `15` characters, as shown in the following example:
```

| :--- | :--- | :--- | :--- |
| `field` | yes | - | The field to get base64 encoded binary from. |
| `target_field` | no | attachment | The field that holds the attachment information. |
| `properties` | no | all properties | An array of properties, which should be stored. Can be `content`, `language`, `date`, `title`, `author`, `keywords`, `content_type`, `content_length`. |
| `indexed_chars` | no | `100_000` | The number of character used for extraction to prevent fields from becoming to large. Use `-1` for no limit. |
| `indexed_chars_field` | no | `null` | The field name from which you can overwrite the number of chars being used for extraction, for example, `indexed_chars`. |
| `ignore_missing` | no | `false` | When `true`, the processor exits without modifying the document when the specified field doesn't exist. |

## Example

The following steps show how to get started with the Ingest-attachment plugin.

### Create an index to store your attachments

The following command creates an index for storing your attachments:

```json
PUT /example-attachment-index
{
  "mappings": {
    "properties": {}
  }
}
```

### Create a pipeline with attachment processor

> The following command creates a pipeline wihch contains the attachment processor:

Reviewer 2: ```suggestion
The following command creates a pipeline which contains the attachment processor:
```

---
layout: default
title: Ingest-attachment plugin
parent: Installing plugins
nav_order: 20

---

# Ingest-attachment plugin

The `ingest-attachment` plugin enables OpenSearch to extract content and other
information from files using the Apache text extraction library
[Tika](https://tika.apache.org/).

> Supported document formats include PPT, PDF, RTF, ODF and many more.

Reviewer 3: Is there any way to link to tika documentation which lists out supported file formats?

Author: You're right, that is the better approach. I skipped it just out of concerns, as it is dependent on the tika version used here: https://github.com/opensearch-project/OpenSearch/blob/main/plugins/ingest-attachment/build.gradle
Currently this is 2.9.2
I will link to supported formats here: https://tika.apache.org/2.9.2/formats.html

But it might be necessary to keep this from time to time updated, especially, once newer versions of tika are imported - like the upcoming version 3.0.0 (currently beta)

The input field must be a base64 encoded binary.

## Installing the plugin

You can install the `ingest-attachment` plugin using the following command:

```sh
./bin/opensearch-plugin install ingest-attachment
```

## Attachment processor options

| Name | Required | Default | Description |
| :--- | :--- | :--- | :--- |
| field | yes | - | The field to get base64 encoded binary from. |

> | target_field| no | attachment | The field that will hold the attachment information. |

Reviewer 3: nit: needs extra space after `target_field`
